ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java

```diff
@@ -126,131 +130,123 @@ private JceMasterKey createJceMasterKey(String tenantId) {
         return JceMasterKey.getInstance(new SecretKeySpec(bytes, "AES"), "Custom", "", "AES/GCM/NOPADDING");
     }
 
-    private void initMasterKey(String tenantId) {
-        if (tenantMasterKeys.containsKey(Objects.requireNonNullElse(tenantId, DEFAULT_TENANT_ID))) {
+    private void initMasterKey(String tenantId, ActionListener<Boolean> listener) {
+        tenantId = Objects.requireNonNullElse(tenantId, DEFAULT_TENANT_ID);
+        if (tenantMasterKeys.containsKey(tenantId)) {
+            log.info("Master key is already available");
+            listener.onResponse(true);
             return;
         }
-        String masterKeyId = MASTER_KEY;
-        if (tenantId != null) {
-            masterKeyId = MASTER_KEY + "_" + hashString(tenantId);
+        // checking and setting the waiting listeners for master key generation
+        List<ActionListener<Boolean>> newWaitingListenersPerTenant = new ArrayList<>();
+        List<ActionListener<Boolean>> existingWaitingListenersPerTenant = tenantWaitingListenerMap
+            .putIfAbsent(tenantId, newWaitingListenersPerTenant);
+        List<ActionListener<Boolean>> waitingListenersPerTenant = (existingWaitingListenersPerTenant != null)
+            ? existingWaitingListenersPerTenant
+            : newWaitingListenersPerTenant;
+        waitingListenersPerTenant.add(listener);
+        synchronized (waitingListenersPerTenant) {
+            try {
+                // checking and waiting if the master key generation triggered by any other thread
+                if (existingWaitingListenersPerTenant != null && tenantWaitingListenerMap.containsKey(tenantId)) {
+                    log.info("Waiting for other thread to generate master key");
+                    waitingListenersPerTenant.wait();
```

Comment:
This is a blocking call, which we're trying to avoid, and not really an improvement over the CountDownLatch.

What we should be doing is returning from this code without completing the action listener; and when the thread that generates the master key returns, it should complete all waiting listeners.